### PR TITLE
Add limit to readAllBytes()/readAllText()

### PR DESCRIPTION
**Pull request by harrishancock**
Reading an unbounded amount of data from a stream is a potential DoS vector. To manage this risk, `readAllText()` and `readAllBytes()` now accept a `limit` parameter. For source backwards-compatibility, this limit defaults to `kj::maxValue`.

The presence of this limit might make it seem that `readAllText()`/`readAllBytes()` are now misnamed. If we care, we could add `read*Until()` overloads that accept a mandatory parameter, and deprecate the `readAll*()` versions.

### CODE REVIEW

File: c++/src/kj/async-io-test.c++ [hunk 1]
```diff
@@ -681,6 +681,52 @@ kj::Promise<void> expectRead(kj::AsyncInputStream& in, kj::StringPtr expected) {
...
+  WaitScope ws(loop);
+
+  auto bigText = strArray(kj::repeat("foo bar baz"_kj, 12345), ",");
+  size_t blockSizes[] = { 1, 4, 256, 4096, 8192, bigText.size() };
+
+  for (size_t blockSize: blockSizes) {
+    {
+      KJ_CONTEXT(blockSize);
+      MockAsyncInputStream input(bigText.asBytes(), blockSize);
+      KJ_EXPECT(input.readAllText().wait(ws) == bigText);
```
Comment (kentonv) [inline c++/src/kj/async-io-test.c++]: Not testing with a limit here?

Seems like we actually want a two-dimensional test matrix, one to set the stream's block size, and one to choose the read limit? And then test readAllText() and readAllBytes() for each?

**Comment by harrishancock** [comment]
Yep. I had meant to call out the exception-versus-return-the-limit design choice explicitly. Given the naming of the function, you're right that throwing an exception does make more sense. Will fix.

File: c++/src/kj/async-io-test.c++ (hunk 1 above)
Comment (harrishancock) [reply c++/src/kj/async-io-test.c++]: That makes sense.

**Comment by harrishancock** [comment]
Updated.

I made it throw an exception if `limit` bytes are read before seeing EOF, rather than throwing an exception if we read more than `limit` bytes. This means `readAllText(0)` will unconditionally throw, and `readAllText(1)` throws if the stream is not at EOF. The alternative requires us to potentially perform an extra read to detect an EOF before finishing the algorithm, or to unconditionally increment the limit by one (making sure to clamp to `kj::maxValue`) before running the algorithm.

Since the only situation I could see where someone would care about the difference in behavior is if they already know in advance how much data they are about to read (and should thus be using `read()` or `tryRead()`), I figured we should prioritize implementation cleanliness.

I also used a three-dimensional test matrix instead of a two-dimensional one: in addition to simulated InputStream block size and read-all limit, I think we need to test when a stream ends at or near a read-all block boundary.

**Comment by harrishancock** [comment]
Cygwin failure looked like a transient connection problem, pushed a new commit to try again.

File: c++/src/kj/io.h [hunk 2]
```diff
@@ -29,6 +29,7 @@
 #include "common.h"
 #include "array.h"
 #include "exception.h"
+#include <cstdint>
```
[the remark this answers is not in the document]
Comment (harrishancock) [reply c++/src/kj/io.h]: Fixed.

> I believe `<cwhatever>` headers are deprecated.

Other way around actually -- `<whatever.h>` headers were deprecated in C++98. You're right about the namespacing issue ... but we can make a similar argument that we should be wrapping every .h in an `extern "C"` block. Both choices are left to the implementation. :(
